packages/wallet/src/Wallet.d.ts

```diff
@@ -0,0 +1,308 @@
+import type {
+  AccountsController,
+  AccountsControllerActions,
+  AccountsControllerEvents,
+  AccountsControllerState,
+} from '@metamask/accounts-controller';
+import type {
+  ApprovalController,
+  ApprovalControllerActions,
+  ApprovalControllerEvents,
+  ApprovalControllerState,
+} from '@metamask/approval-controller';
+import type {
+  CurrencyRateController,
+  CurrencyRateControllerActions,
+  CurrencyRateControllerEvents,
+  CurrencyRateState,
+} from '@metamask/assets-controllers';
+import type {
+  ControllerGetStateAction,
+  ControllerMessenger,
+  ControllerStateChangeEvent,
+} from '@metamask/base-controller';
+import type {
+  GasFeeController,
+  GasFeeControllerActions,
+  GasFeeControllerEvents,
+  GasFeeState,
+} from '@metamask/gas-fee-controller';
+import type { JsonRpcEngine } from '@metamask/json-rpc-engine';
+import type {
+  KeyringController,
+  KeyringControllerActions,
+  KeyringControllerEvents,
+  KeyringControllerState,
+} from '@metamask/keyring-controller';
+import type {
+  NetworkController,
+  NetworkControllerActions,
+  NetworkControllerEvents,
+  NetworkState,
+} from '@metamask/network-controller';
+import type {
+  CaveatSpecificationConstraint,
+  ExtractPermission,
+  PermissionController,
+  PermissionControllerActions,
+  PermissionControllerEvents,
+  PermissionControllerState,
+  PermissionSpecificationConstraint,
+  SubjectType,
+} from '@metamask/permission-controller';
+import type {
+  PreferencesController,
+  PreferencesControllerActions,
+  PreferencesControllerEvents,
+  PreferencesState,
+} from '@metamask/preferences-controller';
+import type {
+  TransactionController,
+  TransactionControllerActions,
+  TransactionControllerEvents,
+  TransactionControllerState,
+} from '@metamask/transaction-controller';
+
+import type {
+  InternalCaveatSpecification,
+  InternalPermissionSpecification,
+} from './permissions';
+
+/**
+ * MetaMask wallet state.
+ */
+export type MetamaskState<
+  ControllerPermissionSpecification extends PermissionSpecificationConstraint &
```

Comment:
Making the permissions extensible ends up being quite verbose. Surely there's a more concise way of doing this, but I haven't experimented with it much yet.